regex-syntax/src/parser.rs

```diff
@@ -526,65 +526,132 @@ impl Parser {
 
     // Parses a character class, e.g., `[^a-zA-Z0-9]+`.
     //
+    // If the Unicode flag is enabled, the class is returned as a `CharClass`,
+    // otherwise it is converted to a `ByteClass`.
+    //
     // Start: `[`
     // End:   `+`
     fn parse_class(&mut self) -> Result<Build> {
+        let class = try!(self.parse_class_as_chars());
+        Ok(Build::Expr(if self.flags.unicode {
+            Expr::Class(class)
+        } else {
+            let byte_class = class.to_byte_class();
+
+            // If `class` was only non-empty due to multibyte characters, the
+            // corresponding byte class will now be empty.
+            //
+            // See https://github.com/rust-lang/regex/issues/303
+            if byte_class.is_empty() {
+                // e.g., (?-u)[^\x00-\xFF]
+                return Err(self.err(ErrorKind::EmptyClass));
+            }
+
+            Expr::ClassBytes(byte_class)
+        }))
+    }
+
+    // Parses a character class as a `CharClass`, e.g., `[^a-zA-Z0-9]+`.
+    //
+    // This does not convert to a `ByteClass` yet, so that it can be used for
+    // nested character classes.
+    //
+    // Start: `[`
+    // End:   `+`
+    fn parse_class_as_chars(&mut self) -> Result<CharClass> {
         self.bump();
         let negated = self.bump_if('^');
+
+        let mut class = try!(self.parse_class_set(true));
+        loop {
+            match self.cur() {
+                ']' => {
+                    // end of the class
+                    self.bump();
+                    break;
+                }
+                '&' => {
+                    // intersection with `&&`
+                    self.bump();
+                    self.bump();
```

Comment:
Shouldn't this throw a syntax error (or just treat it as a literal `&` in the character class, not sure what spec says) if the second character you `bump` over isn't also an `&`?

**Edit:** noticed later that you only exit `parse_class_set` if you look ahead and see `&&`. Maybe add a comment on the second bump saying this was verified to be `&` in parse_class_set.